x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/CoordinatorProxyAction.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.enrich.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.search.MultiSearchRequest;
+import org.elasticsearch.action.search.MultiSearchResponse;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.enrich.EnrichPlugin;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * An internal action to locally manage the load of the search requests that originate from the enrich processor.
+ * This is because the enrich processor executes asynchronously and a bulk request could easily overload
+ * the search tp.
+ */
+public class CoordinatorProxyAction extends ActionType<SearchResponse> {
+
+    public static final CoordinatorProxyAction INSTANCE = new CoordinatorProxyAction();
+    public static final String NAME = "indices:data/read/xpack/enrich/coordinate_lookups";
+
+    private CoordinatorProxyAction() {
+        super(NAME, SearchResponse::new);
+    }
+
+    public static class TransportAction extends HandledTransportAction<SearchRequest, SearchResponse> {
+
+        private final Coordinator coordinator;
+
+        @Inject
+        public TransportAction(TransportService transportService, ActionFilters actionFilters, Coordinator coordinator) {
+            super(NAME, transportService, actionFilters, (Writeable.Reader<SearchRequest>) SearchRequest::new);
+            this.coordinator = coordinator;
+        }
+
+        @Override
+        protected void doExecute(Task task, SearchRequest request, ActionListener<SearchResponse> listener) {
+            assert Thread.currentThread().getName().contains(ThreadPool.Names.WRITE);
+            coordinator.schedule(request, listener);
+        }
+    }
+
+    public static class Coordinator {
```

Comment:
Let's have unit tests for this coordinator?

Reply:
yup, already working on it ;)

Reply:
I pushed: d5e3ac16c6029e54c71bcb3b24e17c6b2a10f338